src/pieces.py

```diff
@@ -9,118 +9,272 @@
     Rook: Moves along ranks and files. Worth 5.
     Queen: Moves both like a rook and bishop. Worth 9.
     King: Moves one square in any direction. Worth N/A (you loose the king it's game over).
+
+Variable annotations:
+    n: north
+    w: west
+    s: south
+    e: east
+
+Concatenation of aforementioned annotations means combination of directions (see self-documenting code).
+Trailing '_' means indefinite.
 """
 
 from dataclasses import dataclass
+from enum import Enum
+from itertools import product, cycle
+from typing import Callable, ClassVar, Iterable
+
+from .square import Vector, Square
+
+
+class Color(Enum):
+    """Annotate direction on the board with the corresponding color."""
+
+    white = -1
+    black = +1
 
 
 @dataclass
 class Piece:
-    """A generic chess piece."""
+    """A generic chess piece.
+
+    Attributes:
+        color: One of two colors, black or white. Holds direction of the board.
+        value: How much a piece is worth:
+            1: Pawn
+            3: Knight/Bishop
+            5: Rook
+            9: Queen
+            ∞: King
+
+    A few generic rules applied to all pieces:
+        Pinning: If moving a piece will discover a check on your king, the piece is pinned and cannot move.
+        If your king is checked, no other piece can move except for the ones and with only the moves that resolve the check.
+    """
+
+    north: ClassVar[Vector] = Vector(-1, 0)
+    south: ClassVar[Vector] = Vector(+1, 0)
+
+    west: ClassVar[Vector] = Vector(0, -1)
+    east: ClassVar[Vector] = Vector(0, +1)
+
+#   Straight steps:
+    rizontals: ClassVar[set[Vector]] = set((
+        west,
+        east,
+    ))
+    verticals: ClassVar[set[Vector]] = set((
+        north,
+        south,
+    ))
+    straights: ClassVar[set[Vector]] = rizontals.union(verticals)
+
+#   Diagonal steps:
+    diagonals: ClassVar[set[Vector]] = set((rizontal + vertical for rizontal, vertical in product(
+        rizontals,
+        verticals,
+    )))
+
+    steps: ClassVar[set[Vector]] = set()  # a set to be filled by various subclasses
```

Comment:
:warning: `steps` here is a class variable, but when `Piece` gets subclassed it is redefined to be a class attribute!